### DISCUSSION TITLE
React functional setState 裡的 alert() 會跳出兩次！？

### DISCUSSION THREAD

**Discussion by KW0409** [Lidemy/forum#250] [前端相關問題] [answered]
如題，在做五子棋的時候，做了一個可以返回上一步的按鈕
因為放整個程式碼感覺會有點太多，所以就只簡略放上跟此按鈕有關的程式碼，如下
```js
function GameButton({ setBoardState }) {
  const handleClick = () => {
    setBoardState((prevState) => {
      let currentStep = prevState.stepRound;
      if (!currentStep) {
        alert("沒有上一步了！");
        return prevState;
      }

      const newStep = --currentStep;
      return {
        ...prevState,
        stepRound: newStep,
        roundBlack: newStep % 2 === 0,
      };
    })
  }

  return (
    <button onClick={handleClick}>
      上一步
    </button>
  )
}
```

奇怪的情況是
如果是還沒有下任何棋子的情況下
點擊『上一步』的按鈕(此時 currentStep 為 0)，只會跳出一次 alert()
但是如果今天下了一步棋之後，
點擊『上一步』(直到 currentStep 為 0)的時候，卻會跳出兩次 alert()

想知道為什麼會有這樣的現象發生？
是因為 setState 是非同步嗎？

**Comment by aszx87410** [comment, marked as answer]
其實我也不是很確定，但我猜可能跟 react fiber 有關，就是 react 可能會把工作切成很多小塊，然後每個小塊有可能會被重新執行，而 alert 可能會造成這個現象，可以把 alert 改成 console 看看是否還會有相同狀況

**Reply by KW0409** [reply]
我有試過放 console 在裡面，但是就只有 alert() 會有出現兩次的狀況

**Reply by aszx87410** [reply]
那應該跟 alert 的特性脫不了關係，alert 跳出的時候基本上會 block 整個畫面

**Reply by KW0409** [reply]
原來如此，感謝老師！
